index.bs

```diff
@@ -931,6 +1019,231 @@ steps:
+  : |handle| . {{FileSystemSyncAccessHandle/write()|write}}(|buffer|, { {{FileSystemReadWriteOptions/at}} })
+  :: Writes the content of |buffer| into the file associated with |handle|, optionally at a given offset.
+</div>
+
+// TODO(fivedots): Figure out how to properly check the available storage quota (in this method and others) by passing the right storage shelf.
+// TODO(fivedots): Figure out mechanism to prevent concurrent IO operations e.g. doing write() while a truncate() is executing in parallel.
+<div algorithm>
+The <dfn method for=FileSystemSyncAccessHandle>write(|buffer|, {{FileSystemReadWriteOptions}}: |options|)</dfn> method, when invoked, must run
+these steps:
+
+1. If [=this=].[=[[state]]=] is "`closed`", throw a {{InvalidStateError}}.
+1. Let |writePosition| be |options|.{{FileSystemReadWriteOptions/at}}.
+1. Let |fileContents| be a copy of [=this=].[=[[file]]=]'s [=file entry/binary data=].
+1. Let |oldSize| be |fileContents|'s [=byte sequence/length=].
+1. Let |data| be [=get a copy of the buffer source|a copy of=] |buffer|.
+
+   Note: The objective of this copy is to prevent concurrent modifications of the buffer
+   from being reflected in the file. Depending on the implementation, a copy
+   of non-shared ArrayBuffers may not be needed in a synchronous call like this one.
+   It might also be possible to do a buffer copy at an earlier time or through a
+   different mechanism.
+
+1. If |writePosition| is larger than |oldSize|,
+   append |writePosition| &minus; |oldSize| `0x00` (NUL) bytes to the end of |fileContents|.
+
+   Note: Implementations are expected to behave as if the skipped over file contents
+   are indeed filled with NUL bytes. That doesn't mean these bytes have to actually be
+   written to disk and take up disk space. Instead most file systems support so called
+   sparse files, where these NUL bytes don't take up actual disk space.
+
+1. Let |head| be a [=byte sequence=] containing the first |writePosition| bytes of |fileContents|.
+1. Let |tail| be an empty [=byte sequence=].
+1. If |writePosition| + |data|'s [=byte sequence/length=] is smaller than |oldSize|:
+   1. Set |tail| to a [=byte sequence=] containing the last
+      |oldSize| &minus; (|writePosition| + |data|'s [=byte sequence/length=]) bytes of |fileContents|.
+1. Let |newSize| be |head|'s [=byte sequence/length=] + |data|'s [=byte sequence/length=] + |tail|'s [=byte sequence/length=].
+1. If |newSize| &minus; |oldSize| exceeds the available [=storage quota=], throw a {{QuotaExceededError}}.
+1. Set [=this=].[=[[file]]=]'s [=file entry/binary data=] to the concatenation of |head|, |data| and |tail|.
+1. If the operations modifying the [=this=].[=[[file]]=]'s [=file entry/binary data=] in the previous steps
+   failed:
+    1. If there were partial writes and the number of bytes that were written from |data| is known,
+       return the number of bytes that were written from |data|.
+    1. Else return 0.
```

Comment:
Well, in full disk cases write() will return less than the requested amount, and will have modified (and maybe extended) the file. (on linux)

Reply:
Whoops, I did mean that it throws, since there is no promise to reject!

We've since [updated the Chrome implementation](https://crrev.com/c/3657990) to support partial writes. The write method should only reject when nothing could be written